presto-tests/src/test/java/com/facebook/presto/tests/TestHistoryBasedRetry.java

```diff
@@ -0,0 +1,222 @@
+                .build();
+
+        try (QueryRunner queryRunner = createQueryRunner(session)) {
+            List<ListenableFuture<?>> queryFutures = new ArrayList<>();
+            for (int i = 0; i < totalQueries; i++) {
+                ListenableFuture<?> future = executor.submit(() -> queryRunner.execute(session, sql));
+                queryFutures.add(future);
+            }
+            waitForQueryToFinish(queryFutures);
+
+            DistributedQueryRunner x = (DistributedQueryRunner) queryRunner;
+            List<BasicQueryInfo> queryInfos = x.getCoordinator().getQueryManager().getQueries();
+
+            for (BasicQueryInfo info : queryInfos) {
+                System.out.println(info.getErrorCode());
+                if (info.getQuery().contains("-- retry query")) {
+                    retryNum++;
+                }
+            }
+        }
+
+        assertEquals(retryNum, 1, "Retry count should be one as the query plan has changed.");
+    }
+
+    @Test
+    public void testQueryRetryWithHBONewQueryPlan()
+            throws Exception
+    {
+        int totalQueries = 1;
+        int retryLimit = 3;
+        int retryNum = 0;
+
+        String sql = "select o.orderkey, l.partkey, l.mapcol[o.orderkey] from (select orderkey, partkey, mapcol " +
+                "from (select *, map(array[1], array[2]) mapcol from lineitem)) l " +
+                "join orders o on l.partkey=o.custkey where length(comment)>30";
+
+        Session session = Session.builder(createSession())
+                .setSystemProperty(QUERY_RETRY_LIMIT, String.valueOf(retryLimit))
+                .build();
+
+        try (QueryRunner queryRunner = createQueryRunner(session)) {
+            List<ListenableFuture<?>> queryFutures = new ArrayList<>();
+            for (int i = 0; i < totalQueries; i++) {
+                ListenableFuture<?> future = executor.submit(() -> queryRunner.execute(session, sql));
+                queryFutures.add(future);
+            }
+            waitForQueryToFinish(queryFutures);
+
+            DistributedQueryRunner x = (DistributedQueryRunner) queryRunner;
+            List<BasicQueryInfo> queryInfos = x.getCoordinator().getQueryManager().getQueries();
+
+            for (BasicQueryInfo info : queryInfos) {
+                System.out.println(info.getErrorCode());
+                if (info.getQuery().contains("-- retry query")) {
+                    retryNum++;
+                }
+            }
+        }
+        assertEquals(retryNum, 1, "Retry count should be one as the query plan has changed.");
+    }
+
+    @Test
+    public void testQueryRetryWithHBOForceFail()
+            throws Exception
+    {
+        int totalQueries = 1;
+        int retryLimit = 1;
+        int retryNum = 0;
+
+        String query = "SELECT if(COUNT(*)=1,1,fail(1, 'failed')) FROM part p LEFT JOIN lineitem l ON p.partkey = l.partkey WHERE l.comment like '%a%'";
+
+        Session session = Session.builder(createSession())
```

Comment:
Haven't looked to closely, but a lot of these methods seem to have a lot of duplicate code. It might be possible to reuse some of that, but don't bother if it's too much trouble